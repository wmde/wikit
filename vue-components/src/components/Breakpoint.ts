// eslint-disable-next-line no-shadow
enum Breakpoint {
	Mobile = 'mobile',
	Tablet = 'tablet',
	Desktop = 'desktop'
}

function validateBreakpoint( breakpoint: string ): boolean {
	return Object.values( Breakpoint ).includes( breakpoint as Breakpoint );
}

export {
	Breakpoint,
	validateBreakpoint,
};
