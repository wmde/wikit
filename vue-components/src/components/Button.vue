<template>
	<button
		:class="[
			'wikit',
			'wikit-Button',
			`wikit-Button--${ type }`,
			`wikit-Button--${ variant }`,
			iconOnly ? `wikit-Button--iconOnly` : '',
		]"
	>
		<slot />
	</button>
</template>

<script lang="ts">
import Vue from 'vue';
import VueCompositionAPI, { defineComponent, onMounted } from '@vue/composition-api';

Vue.use( VueCompositionAPI );

/**
 *  An interactive element signaling a single-step action that will occur when the user clicks or taps on it.
 *
 * Known issues:
 *
 * * The styles defined on `:focus` do not apply in Safari and Firefox on macOS.<br>
 * This seems to be a desired behavior and not a bug<br>
 * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus<br>
 * https://bugzilla.mozilla.org/show_bug.cgi?id=1581369#c5
 */
export default defineComponent( {
	name: 'Button',
	props: {
		/**
		 * The type of the button
		 *
		 * Allowed values: `neutral`, `progressive`, `destructive`
		 */
		type: {
			type: String,
			validator( value: string ): boolean {
				return [ 'neutral', 'progressive', 'destructive' ].includes( value );
			},
			default: 'neutral',
		},
		/**
		 * The variant of the button
		 *
		 * Allowed values: `normal`, `primary`
		 */
		variant: {
			type: String,
			validator( value: string ): boolean {
				return [ 'normal', 'primary' ].includes( value );
			},
			default: 'normal',
		},
		iconOnly: {
			type: Boolean,
			default: false,
		},
	},
	setup( props: {
		type: 'neutral'|'progressive'|'destructive';
		variant: 'normal'|'primary';
	} ) {
		onMounted( () => {
			const supportedCombinations = {
				normal: [ 'neutral' ],
				primary: [ 'progressive', 'destructive' ],
			};
			if ( !supportedCombinations[ props.variant ].includes( props.type ) ) {
				throw new Error(
					`The combination of variant "${props.variant}" and type "${props.type}" is not yet supported!`,
				);
			}
		} );
	},
} );
</script>

<style lang="scss">
$base: '.wikit-Button';

#{$base} {
	padding-inline: $wikit-Button-medium-padding-horizontal;
	padding-block: $wikit-Button-medium-padding-vertical;
	font-family: $wikit-Button-font-family;
	font-weight: $wikit-Button-font-weight;
	font-size: $wikit-Button-font-size; // unit? px vs em vs rem
	line-height: $wikit-Button-line-height;
	cursor: pointer;
	border-width: $wikit-Button-border-width;
	border-style: $wikit-Button-border-style;
	border-radius: $wikit-Button-border-radius;
	transition-duration: $wikit-Button-transition-duration;
	transition-timing-function: $wikit-Button-transition-timing-function;
	transition-property: $wikit-Button-transition-property;
	white-space: nowrap;

	// TODO use breakpoint mixin?
	@media (max-width: $width-breakpoint-mobile) {
		padding-inline: $wikit-Button-large-padding-horizontal;
		padding-block: $wikit-Button-large-padding-vertical;
	}

	&#{$base}--normal {
		&#{$base}--neutral {
			color: $wikit-Button-normal-neutral-color;
			background-color: $wikit-Button-normal-neutral-background-color;
			border-color: $wikit-Button-normal-neutral-border-color;

			&:hover {
				color: $wikit-Button-normal-neutral-hover-color;
				background-color: $wikit-Button-normal-neutral-hover-background-color;
				border-color: $wikit-Button-normal-neutral-hover-border-color;
			}

			&:active {
				color: $wikit-Button-normal-neutral-active-color;
				background-color: $wikit-Button-normal-neutral-active-background-color;
				border-color: $wikit-Button-normal-neutral-active-border-color;
			}
			// A clicked button is both :active and :focused. Using :not(:active) to avoid mixing the two.
			&:focus:not(:active) {
				color: $wikit-Button-normal-neutral-focus-color;
				background-color: $wikit-Button-normal-neutral-focus-background-color;
				border-color: $wikit-Button-normal-neutral-focus-border-color;
				box-shadow: $wikit-Button-normal-neutral-focus-box-shadow;
			}
		}
	}

	&#{$base}--primary {
		&#{$base}--progressive {
			color: $wikit-Button-primary-color;
			background-color: $wikit-Button-primary-progressive-background-color;
			border-color: $wikit-Button-primary-progressive-border-color;

			&:hover {
				background-color: $wikit-Button-primary-progressive-hover-background-color;
				border-color: $wikit-Button-primary-progressive-hover-border-color;
			}

			&:active {
				background-color: $wikit-Button-primary-progressive-active-background-color;
				border-color: $wikit-Button-primary-progressive-active-border-color;
			}

			// A clicked button is both :active and :focused. Using :not(:active) to avoid mixing the two.
			&:focus:not(:active) {
				background-color: $wikit-Button-primary-progressive-focus-background-color;
				border-color: $wikit-Button-primary-progressive-focus-border-color;
				box-shadow: $wikit-Button-primary-progressive-focus-box-shadow;
			}
		}

		&#{$base}--destructive {
			color: $wikit-Button-primary-color;
			background-color: $wikit-Button-primary-destructive-background-color;
			border-color: $wikit-Button-primary-destructive-border-color;

			&:hover {
				background-color: $wikit-Button-primary-destructive-hover-background-color;
				border-color: $wikit-Button-primary-destructive-hover-border-color;
			}

			&:active {
				background-color: $wikit-Button-primary-destructive-active-background-color;
				border-color: $wikit-Button-primary-destructive-active-border-color;
			}

			// A clicked button is both :active and :focused. Using :not(:active) to avoid mixing the two.
			&:focus:not(:active) {
				background-color: $wikit-Button-primary-destructive-focus-background-color;
				border-color: $wikit-Button-primary-destructive-focus-border-color;
				box-shadow: $wikit-Button-primary-destructive-focus-box-shadow;
			}
		}
	}

	&#{$base}--normal:disabled,
	&#{$base}--primary:disabled {
		color: $wikit-Button-normal-disabled-color;
		background-color: $wikit-Button-normal-disabled-background-color;
		border-color: $wikit-Button-normal-disabled-border-color;
		cursor: default;
		pointer-events: none;
	}

	&#{$base}--iconOnly {
		padding-inline: $wikit-Button-icon-only-medium-padding-horizontal;

		@media (max-width: $width-breakpoint-mobile) {
			padding-inline: $wikit-Button-icon-only-large-padding-horizontal;
		}
	}

	// should ideally be taken care of by the globally applied style reset (ress)
	// https://github.com/filipelinhares/ress/pull/24
	&:focus {
		outline: none;
	}
}
</style>
