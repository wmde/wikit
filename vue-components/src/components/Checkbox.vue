<template>
	<span class="wikit wikit-checkbox">
		<input
			type="checkbox"
			:id="id"
			class="wikit-checkbox__input"
			@change="$emit( 'update:checked', $event.target.checked )"
			:value="value"
			:checked="checked"
			:disabled="disabled"
		>
		<label class="wikit-checkbox__label" :for="id">{{ label }}</label>
		<slot name="suffix" />
	</span>
</template>

<script lang="ts">
import Vue from 'vue';
import generateUid from '@/components/util/generateUid';

export default Vue.extend( {
	name: 'Checkbox',
	props: {
		checked: {
			required: true,
			type: Boolean,
		},
		value: {
			default: null,
			type: String,
		},
		label: {
			required: true,
			type: String,
		},
		disabled: {
			type: Boolean,
			default: false,
		},
	},
	data() {
		return {
			// https://github.com/vuejs/vue/issues/5886
			id: generateUid( 'wikit-Checkbox' ),
		};
	},
} );
</script>

<style lang="scss">
$label: '.wikit-checkbox__label';

.wikit-checkbox {
	display: flex;

	&__label {
		@include Label;
		display: flex;
		align-items: center;
		margin-inline-end: $dimension-spacing-small;

		&::before {
			min-width: $wikit-Checkbox-input-size;
			min-height: $wikit-Checkbox-input-size;
			box-sizing: border-box;

			border-width: $wikit-Checkbox-input-border-width;
			border-color: $wikit-Checkbox-input-border-color;
			border-style: $wikit-Checkbox-input-border-style;
			border-radius: $wikit-Checkbox-input-border-radius;
			background-color: $wikit-Checkbox-input-background-color; /* Workaround Firefox e.a. */

			content: ' ';
			margin-inline-end: $wikit-Checkbox-input-right-spacing;
			display: flex;
			align-self: flex-start;
			transition-property: $wikit-Checkbox-input-transition-property;
			transition-duration: $wikit-Checkbox-input-transition-duration;
			transition-timing-function: $wikit-Checkbox-input-transition-timing-function;
		}

		&:hover {
			&::before {
				border-color: $wikit-Checkbox-input-hover-border-color;
			}
		}

		&:active {
			&::before {
				border-color: $wikit-Checkbox-input-active-border-color;
				background-color: $wikit-Checkbox-input-active-background-color;
			}
		}
	}

	&__input {/* see: https://www.w3schools.com/howto/howto_css_custom_checkbox.asp */
		width: 0;
		height: 0;
		opacity: 0;
		inset-inline-start: 0;/* we move it out of our way */
		position: absolute;
		&:focus + #{$label} {
			&::before {
				border-color: $wikit-Checkbox-input-focus-border-color;
				box-shadow: $wikit-Checkbox-input-focus-box-shadow;
			}
		}

		&:disabled {
			+ #{$label} {
				color: $wikit-Checkbox-label-disabled-color;
				pointer-events: none;
			}
			&:checked + #{$label}::before {
				border-color: $wikit-Checkbox-input-disabled-border-color;
				background-color: $wikit-Checkbox-input-disabled-background-color;
			}
		}

		&:checked + #{$label} {
			&::before {
				border-color: $wikit-Checkbox-input-selected-border-color;
				background-color: $wikit-Checkbox-input-selected-background-color;
				background-image: svg-load('check.svg', fill=$wikit-Checkbox-input-selected-icon-color);
				background-position: 50% 50%;
				background-repeat: no-repeat;
				background-size: $wikit-Checkbox-input-selected-icon-size;
			}

			&:hover::before {
				background-color: $wikit-Checkbox-input-selected-hover-background-color;
				border-color: $wikit-Checkbox-input-selected-hover-border-color;
			}

			&:active::before {
				background-color: $wikit-Checkbox-input-selected-active-background-color;
				border-color: $wikit-Checkbox-input-selected-active-border-color;
			}
		}
		&:checked:focus + #{$label} {
			&::before {
				box-shadow: $wikit-Checkbox-input-selected-focus-box-shadow;
				background-color: $wikit-Checkbox-input-selected-focus-background-color;
				border-color: $wikit-Checkbox-input-selected-focus-border-color;
			}
		}
	}
}
</style>
