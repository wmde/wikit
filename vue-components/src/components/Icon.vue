<template>
	<!-- eslint-disable max-len -->
	<span :class="['wikit', 'wikit-Icon', `wikit-Icon--${size}`, `wikit-Icon--${color}`]">
		<!-- error icon -->
		<svg
			class="wikit-Icon__svg"
			viewBox="0 0 16 16"
			fill="none"
			xmlns="http://www.w3.org/2000/svg"
			v-if="type === IconTypes.ERROR"
		>
			<path
				fill="currentColor"
				d="M11.4939 0H4.50612L0 4.50612V11.5102L4.50612 16H11.5102L16 11.4939V4.50612L11.4939 0Z
					M8.8 11.9938H7.2V10.3999H8.8V11.9938Z
					M8.8 9.6H7.2V4H8.8V9.6Z"
			/>
		</svg>

		<!-- warning icon -->
		<svg
			class="wikit-Icon__svg"
			viewBox="0 0 16 16"
			fill="none"
			xmlns="http://www.w3.org/2000/svg"
			v-else-if="type === IconTypes.ALERT"
		>
			<path fill="currentColor" d="M9.163 1.68234C9.06078 1.4381 8.89901 1.22746 8.69449 1.07231C8.48997 0.917151 8.25017 0.823144 7.99999 0.800049C7.75116 0.82453 7.51294 0.919178 7.30987 1.07425C7.10679 1.22933 6.94619 1.43922 6.84459 1.68234L0.672272 13.0631C0.0337565 14.2368 0.558251 15.2 1.82768 15.2H14.1723C15.4417 15.2 15.9662 14.2368 15.3277 13.0631L9.163 1.68234ZM8.76013 12.7717H7.23986V11.1528H8.76013V12.7717ZM8.76013 9.53394H7.23986V4.67728H8.76013V9.53394Z" />
		</svg>

		<!-- info icon -->
		<svg
			class="wikit-Icon__svg"
			viewBox="0 0 20 20"
			fill="none"
			xmlns="http://www.w3.org/2000/svg"
			v-else-if="type === IconTypes.INFO"
		>
			<path fill="currentColor" d="M10 0C4.477 0 0 4.477 0 10C0 15.523 4.477 20 10 20C15.523 20 20 15.523 20 10C20 4.477 15.523 0 10 0ZM9 5H11V7H9V5ZM9 9H11V15H9V9Z" />
		</svg>

		<!-- checkmark icon -->
		<svg
			class="wikit-Icon__svg"
			viewBox="0 0 20 20"
			fill="none"
			xmlns="http://www.w3.org/2000/svg"
			v-else-if="type === IconTypes.CHECKMARK"
		>
			<path fill="currentColor" d="M6.34812 14.6259L1.6041 9.65425L0 11.3353L6.34812 18L20 3.693L18.3959 2L6.34812 14.6259Z" />
		</svg>

		<!-- clear icon -->
		<svg
			class="wikit-Icon__svg"
			viewBox="0 0 20 20"
			fill="none"
			xmlns="http://www.w3.org/2000/svg"
			v-else-if="type === IconTypes.CLEAR"
		>
			<path fill="currentColor" d="M10 0a10 10 0 1010 10A10 10 0 0010 0zm5.66 14.24l-1.41 1.41L10 11.41l-4.24 4.25-1.42-1.42L8.59 10 4.34 5.76l1.42-1.42L10 8.59l4.24-4.24 1.41 1.41L11.41 10z" />
		</svg>

		<!-- edit icon -->
		<svg
			xmlns="http://www.w3.org/2000/svg"
			viewBox="0 0 20 20"
			class="wikit-Icon__svg"
			fill="none"
			v-else-if="type === IconTypes.EDIT"
		>
			<path fill="currentColor" d="M16.77 8l1.94-2a1 1 0 000-1.41l-3.34-3.3a1 1 0 00-1.41 0L12 3.23zM1 14.25V19h4.75l9.96-9.96-4.75-4.75z" />
		</svg>
		<!-- trash icon -->
		<svg
			class="wikit-Icon__svg"
			viewBox="0 0 20 20"
			fill="none"
			xmlns="http://www.w3.org/2000/svg"
			v-else-if="type === IconTypes.TRASH"
		>
			<path d="M17 2H13.5L12.5 1H7.5L6.5 2H3V4H17V2ZM4 17C4 17.5304 4.21071 18.0391 4.58579 18.4142C4.96086 18.7893 5.46957 19 6 19H14C14.5304 19 15.0391 18.7893 15.4142 18.4142C15.7893 18.0391 16 17.5304 16 17V5H4V17Z" fill="currentColor" />
		</svg>

		<!-- info icon -->
		<svg
			viewBox="0 0 20 21"
			fill="none"
			xmlns="http://www.w3.org/2000/svg"
			class="wikit-Icon__svg"
			v-else-if="type === IconTypes.INFOOUTLINED"
		>
			<g clip-path="url(#clip0)">
				<path d="M10 19.1667C7.71324 19.1289 5.53071 18.2037 3.9135 16.5865C2.2963 14.9693 1.37109 12.7868 1.33334 10.5C1.37109 8.21324 2.2963 6.0307 3.9135 4.4135C5.53071 2.79629 7.71324 1.87108 10 1.83333C12.2868 1.87108 14.4693 2.79629 16.0865 4.4135C17.7037 6.0307 18.6289 8.21324 18.6667 10.5C18.6223 12.7846 17.695 14.9634 16.0792 16.5792C14.4634 18.195 12.2847 19.1223 10 19.1667ZM10 0.5C8.02219 0.5 6.08879 1.08649 4.4443 2.1853C2.79981 3.28412 1.51809 4.8459 0.761209 6.67316C0.00433286 8.50042 -0.1937 10.5111 0.192152 12.4509C0.578004 14.3907 1.53041 16.1725 2.92894 17.5711C4.32746 18.9696 6.10929 19.922 8.0491 20.3078C9.98891 20.6937 11.9996 20.4957 13.8268 19.7388C15.6541 18.9819 17.2159 17.7002 18.3147 16.0557C19.4135 14.4112 20 12.4778 20 10.5C20 7.84783 18.9464 5.3043 17.0711 3.42893C15.1957 1.55357 12.6522 0.5 10 0.5ZM10.6667 8.5V13.94H12V15.1667H8.09334V13.94H9.33334V9.83333H8V8.5H10.6667ZM9.33334 5.83333H10.6667V7.16667H9.33334V5.83333Z" fill="currentColor" />
			</g>
			<defs>
				<clipPath id="clip0">
					<rect
						width="20"
						height="20"
						fill="transparent"
						transform="translate(0 0.5)"
					/>
				</clipPath>
			</defs>
		</svg>

	</span>
	<!-- eslint-enable max-len -->
</template>

<script lang="ts">
import Vue from 'vue';
import { IconTypes, iconColors, iconSizes } from './iconProps';

/**
 * Every SVG icon that is added to this component needs to have at least one element (e.g. a `<path>`) with
 * `fill="currentColor"` for the part of the icon that is intended to be colored by the `color` prop.
 *
 * More info on `fill="currentColor"`:
 * https://css-tricks.com/cascading-svg-fill-color/
 *
 */
export default Vue.extend( {
	name: 'Icon',
	data: () => { return { IconTypes }; },

	props: {
		type: {
			type: String,
			validator( value: string ): boolean {
				return Object.values( IconTypes ).includes( value as IconTypes );
			},
			required: true,
		},

		color: {
			type: String,
			validator( value: string ): boolean {
				return iconColors.includes( value );
			},
			default: 'base',
		},

		size: {
			type: String,
			validator( value: string ): boolean {
				return iconSizes.includes( value );
			},
			default: 'large',
		},
	},
} );
</script>

<style lang="scss">
.wikit-Icon {
	display: inline-flex;
	align-items: center;

	&--base {
		color: $wikit-Icon-color-base;
	}

	&--error {
		color: $wikit-Icon-color-error;
	}

	&--warning {
		color: $wikit-Icon-color-warning;
	}

	&--notice {
		color: $wikit-Icon-color-notice;
	}

	&--success {
		color: $wikit-Icon-color-success;
	}

	&--large &__svg {
		width: $wikit-Icon-size-large;
		height: $wikit-Icon-size-large;
	}

	&--medium &__svg {
		width: $wikit-Icon-size-medium;
		height: $wikit-Icon-size-medium;
	}
}
</style>
