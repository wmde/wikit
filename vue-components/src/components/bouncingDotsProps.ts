export const bouncingDotsSizes = [ 'small', 'medium' ];
