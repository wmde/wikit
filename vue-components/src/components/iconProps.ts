export enum IconTypes {
	ALERT = 'alert',
	ERROR = 'error',
	INFO = 'info',
	NEWWINDOW = 'newwindow',
	CHECKMARK = 'checkmark',
	CLEAR = 'clear',
	EDIT = 'edit',
	TRASH = 'trash',
	INFOOUTLINED = 'info-outlined',
	LINK = 'link',
	ADD = 'add',
	SEARCH = 'search'
}

export const iconColors = [
	'base',
	'subtle',
	'emphasized',
	'inverted',
	'warning',
	'error',
	'notice',
	'success',
	'inherit',
];

export const iconSizes = [ 'xsmall', 'small', 'medium', 'large' ];
