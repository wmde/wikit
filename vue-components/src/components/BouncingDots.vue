<template>
	<div
		class="wikit wikit-bouncingdots"
		:class="[ `wikit-bouncingdots`, `wikit-bouncingdots--${size}` ]"
	>
		<div class="bounce" />
	</div>
</template>

<script lang="ts">
import Vue from 'vue';
import { bouncingDotsSizes } from './bouncingDotsProps';

export default Vue.extend( {
	name: 'BouncingDots',
	props: {
		size: {
			type: String,
			validator( value: string ): boolean {
				return bouncingDotsSizes.includes( value );
			},
			default: 'small',
		},
	},
} );
</script>

<style lang="scss">

.wikit-bouncingdots {
	white-space: nowrap;

	.bounce,
	&::before,
	&::after {
		content: '';
		display: inline-block;
		background-color: $wikit-BouncingDots-background-color;
		border-radius: $wikit-BouncingDots-border-radius;
		animation: bounce-delay $wikit-BouncingDots-animation-duration infinite ease-in-out both;
		animation-delay: $wikit-BouncingDots-animation-delay-default;
	}

	&::before {
		margin-inline-end: $wikit-BouncingDots-margin-right;
		animation-delay: $wikit-BouncingDots-animation-delay-before;
	}

	&::after {
		margin-inline-start: $wikit-BouncingDots-margin-left;
		animation-delay: $wikit-BouncingDots-animation-delay-after;
	}

	&--small .bounce,
	&--small::before,
	&--small::after {
		width: $wikit-BouncingDots-width-small;
		height: $wikit-BouncingDots-width-small;
	}

	&--medium .bounce,
	&--medium::before,
	&--medium::after {
		width: $wikit-BouncingDots-width-medium;
		height: $wikit-BouncingDots-width-medium;
	}
}

@keyframes bounce-delay {
	0%,
	50%,
	100% {
		transform: scale(0.625);
	}

	20% {
		opacity: 0.87;
		transform: scale(1);
	}
}
</style>
