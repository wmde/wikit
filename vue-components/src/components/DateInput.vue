<template>
	<div
		:class="[ 'wikit', 'wikit-DateInput' ]"
	>
		<InputWithExtender
			:label="label"
			:value="value"
			:error="error"
			:placeholder="placeholder"
			:disabled="disabled"
			@input="$emit( 'input', $event )"
		>
			<template v-slot:default>
				<p class="wikit-DateInput__IntroText">
					{{ resultsIntroText }}
				</p>
				<p class="wikit-DateInput__ParsedValue" v-if="parsedValue">
					<b>{{ parsedValue }}</b>
				</p>
				<p class="wikit-DateInput__Prompt" v-if="promptText && !value">
					{{ promptText }}
				</p>
				<BouncingDots v-if="!parsedValue && value" type="small" />
				<div class="wikit-DateInput__CalendarNotice" v-if="calendarNotice && parsedValue">
					<Icon color="inherit" type="info" size="small" />
					{{ calendarNotice }}
				</div>
			</template>
			<template v-slot:suffix>
				<slot name="suffix" />
			</template>
		</InputWithExtender>
	</div>
</template>

<script lang="ts">
import Vue from 'vue';
import VueCompositionAPI, { defineComponent } from '@vue/composition-api';
import { errorProp } from '@/compositions/validatable';
import Icon from '@/components/Icon.vue';
import InputWithExtender from '@/components/InputWithExtender.vue';
import BouncingDots from '@/components/BouncingDots.vue';

Vue.use( VueCompositionAPI );

export default defineComponent( {
	name: 'DateInput',
	components: {
		BouncingDots,
		Icon,
		InputWithExtender,
	},
	props: {
		error: errorProp,
		disabled: {
			type: Boolean,
			default: false,
		},
		placeholder: {
			type: String,
			default: '',
		},
		label: {
			type: String,
			default: null,
		},
		parsedValue: {
			type: String,
			default: null,
		},
		resultsIntroText: {
			type: String,
			default: null,
		},
		promptText: {
			type: String,
			default: null,
		},
		value: {
			type: String,
			default: null,
		},
		calendarNotice: {
			type: String,
			default: null,
		},
	},
} );
</script>

<style lang="scss">
.wikit-DateInput {
	&__IntroText {
		padding-block-end: $wikit-InputExtender-content-padding-vertical;
		font-family: $font-family-style-label;
		font-weight: $font-weight-style-label;
		font-size: $font-size-style-label;
		line-height: $font-line-height-style-label;
		color: $font-color-base;
	}

	&__ParsedValue {
		padding-block: $wikit-InputExtender-content-padding-vertical;
		font-family: $font-family-style-label;
		font-weight: $font-weight-bold;
		font-size: $font-size-style-label;
		line-height: $font-line-height-style-label;
		color: $font-color-base;
	}

	&__CalendarNotice {
		padding-block-start: $wikit-InputExtender-content-padding-vertical;
		font-family: $font-family-style-description;
		font-weight: $font-weight-style-description;
		font-size: $font-size-style-description;
		line-height: $font-line-height-style-description;
		color: $font-color-subtle;
	}

	&__Prompt {
		font-family: $font-family-style-placeholder;
		font-size: $font-size-style-placeholder;
		font-weight: $font-weight-style-placeholder;
		line-height: $font-line-height-style-placeholder;
		color: $font-color-subtle;
	}
}
</style>
