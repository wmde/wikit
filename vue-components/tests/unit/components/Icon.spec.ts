import { mount } from '@vue/test-utils';
import Icon from '@/components/Icon.vue';
import { iconSizes, iconColors, IconTypes } from '@/components/iconProps';

describe( 'Icon', () => {

	it( 'renders an error icon given type "error"', () => {
		const wrapper = mount( Icon, {
			propsData: {
				type: 'error',
			},
		} );

		expect( wrapper.element ).toMatchSnapshot();
	} );

	it( 'renders an alert icon given type "alert"', () => {
		const wrapper = mount( Icon, {
			propsData: {
				type: 'alert',
			},
		} );

		expect( wrapper.element ).toMatchSnapshot();
	} );

	it( 'renders an edit icon given type "edit"', () => {
		const wrapper = mount( Icon, {
			propsData: {
				type: 'edit',
			},
		} );

		expect( wrapper.element ).toMatchSnapshot();
	} );

	it( 'renders an error icon given type "trash"', () => {
		const wrapper = mount( Icon, {
			propsData: {
				type: 'trash',
			},
		} );

		expect( wrapper.element ).toMatchSnapshot();
	} );

	it( 'renders a clear icon given type "clear"', () => {
		const wrapper = mount( Icon, {
			propsData: {
				type: 'clear',
			},
		} );

		expect( wrapper.element ).toMatchSnapshot();
	} );

	it( 'validates the type prop', () => {
		expect( () => mount( Icon, {
			propsData: {
				type: 'potato',
			},
		} ) ).toThrow();
	} );

	it( 'validates the color prop', () => {
		expect( () => mount( Icon, {
			propsData: {
				type: 'error',
				color: 'potato',
			},
		} ) ).toThrow();
	} );

	it( 'validates the size prop', () => {
		expect( () => mount( Icon, {
			propsData: {
				type: 'error',
				size: 'potato',
			},
		} ) ).toThrow();
	} );

	it.each(
		Object.values( IconTypes ),
	)( '%s: has at least one part of the SVG element that can be colored', ( iconType ) => {
		const wrapper = mount( Icon, {
			propsData: {
				type: iconType,
			},
		} );

		expect( wrapper.find( '[fill="currentColor"]' ).exists() ).toBeTruthy();
	} );

	it.each( iconColors )( 'renders the color "%s" as a root node class', ( color ) => {
		expect( mount( Icon, {
			propsData: {
				type: 'alert',
				color,
			},
		} ).classes() ).toContain( `wikit-Icon--${color}` );
	} );

	it.each( iconSizes )( 'renders the size "%s" as a root node class', ( size ) => {
		expect( mount( Icon, {
			propsData: {
				type: 'alert',
				size,
			},
		} ).classes() ).toContain( `wikit-Icon--${size}` );
	} );

} );
