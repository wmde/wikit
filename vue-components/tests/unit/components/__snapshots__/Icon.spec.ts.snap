// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders a clear icon given type "clear" 1`] = `
<span
  class="wikit wikit-Icon wikit-Icon--large wikit-Icon--base"
>
  <svg
    class="wikit-Icon__svg"
    fill="none"
    viewBox="0 0 20 20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M10 0a10 10 0 1010 10A10 10 0 0010 0zm5.66 14.24l-1.41 1.41L10 11.41l-4.24 4.25-1.42-1.42L8.59 10 4.34 5.76l1.42-1.42L10 8.59l4.24-4.24 1.41 1.41L11.41 10z"
      fill="currentColor"
    />
  </svg>
</span>
`;

exports[`Icon renders an alert icon given type "alert" 1`] = `
<span
  class="wikit wikit-Icon wikit-Icon--large wikit-Icon--base"
>
  <svg
    class="wikit-Icon__svg"
    fill="none"
    viewBox="0 0 16 16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M9.163 1.68234C9.06078 1.4381 8.89901 1.22746 8.69449 1.07231C8.48997 0.917151 8.25017 0.823144 7.99999 0.800049C7.75116 0.82453 7.51294 0.919178 7.30987 1.07425C7.10679 1.22933 6.94619 1.43922 6.84459 1.68234L0.672272 13.0631C0.0337565 14.2368 0.558251 15.2 1.82768 15.2H14.1723C15.4417 15.2 15.9662 14.2368 15.3277 13.0631L9.163 1.68234ZM8.76013 12.7717H7.23986V11.1528H8.76013V12.7717ZM8.76013 9.53394H7.23986V4.67728H8.76013V9.53394Z"
      fill="currentColor"
    />
  </svg>
</span>
`;

exports[`Icon renders an error icon given type "error" 1`] = `
<span
  class="wikit wikit-Icon wikit-Icon--large wikit-Icon--base"
>
  <svg
    class="wikit-Icon__svg"
    fill="none"
    viewBox="0 0 16 16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M11.4939 0H4.50612L0 4.50612V11.5102L4.50612 16H11.5102L16 11.4939V4.50612L11.4939 0Z
				M8.8 11.9938H7.2V10.3999H8.8V11.9938Z
				M8.8 9.6H7.2V4H8.8V9.6Z"
      fill="currentColor"
    />
  </svg>
</span>
`;
